src/latexify/codegen/expression_codegen.py

```diff
@@ -0,0 +1,704 @@
+        """
+        if not operand_rule.wrap:
+            return self.visit(child)
+
+        if isinstance(child, ast.Call):
+            child_fn_name = ast_utils.extract_function_name_or_none(child)
+            rule = (
+                constants.BUILTIN_FUNCS.get(child_fn_name)
+                if child_fn_name is not None
+                else None
+            )
+            if rule is not None and rule.is_wrapped:
+                return self.visit(child)
+
+        if not isinstance(child, ast.BinOp):
+            return self._wrap_operand(child, parent_prec)
+
+        latex = self.visit(child)
+
+        if _BIN_OP_RULES[type(child.op)].is_wrapped:
+            return latex
+
+        child_prec = _get_precedence(child)
+
+        if child_prec > parent_prec or (
+            child_prec == parent_prec and not operand_rule.force
+        ):
+            return latex
+
+        return rf"\mathopen{{}}\left( {latex} \mathclose{{}}\right)"
+
+    def visit_BinOp(self, node: ast.BinOp) -> str:
+        """Visit a BinOp node."""
+        prec = _get_precedence(node)
+        rule = self._bin_op_rules[type(node.op)]
+        lhs = self._wrap_binop_operand(node.left, prec, rule.operand_left)
+        rhs = self._wrap_binop_operand(node.right, prec, rule.operand_right)
+        return f"{rule.latex_left}{lhs}{rule.latex_middle}{rhs}{rule.latex_right}"
+
+    def visit_UnaryOp(self, node: ast.UnaryOp) -> str:
+        """Visit a unary op node."""
+        latex = self._wrap_operand(node.operand, _get_precedence(node))
+        return _UNARY_OPS[type(node.op)] + latex
+
+    def visit_Compare(self, node: ast.Compare) -> str:
+        """Visit a compare node."""
+        parent_prec = _get_precedence(node)
+        lhs = self._wrap_operand(node.left, parent_prec)
+        ops = [self._compare_ops[type(x)] for x in node.ops]
+        rhs = [self._wrap_operand(x, parent_prec) for x in node.comparators]
+        ops_rhs = [f" {o} {r}" for o, r in zip(ops, rhs)]
+        return lhs + "".join(ops_rhs)
+
+    def visit_BoolOp(self, node: ast.BoolOp) -> str:
+        """Visit a BoolOp node."""
+        parent_prec = _get_precedence(node)
+        values = [self._wrap_operand(x, parent_prec) for x in node.values]
+        op = f" {_BOOL_OPS[type(node.op)]} "
+        return op.join(values)
+
+    def visit_IfExp(self, node: ast.IfExp) -> str:
+        """Visit an ifexp node"""
+        latex = r"\left\{ \begin{array}{ll} "
+
+        current_expr: ast.expr = node
+
+        while isinstance(current_expr, ast.IfExp):
+            cond_latex = self.visit(current_expr.test)
+            true_latex = self.visit(current_expr.body)
+            latex += true_latex + r", & \mathrm{if} \ " + cond_latex + r" \\ "
+            current_expr = current_expr.orelse
+
+        latex += self.visit(current_expr)
+        return latex + r", & \mathrm{otherwise} \end{array} \right."
+
+    def _reduce_stop_parameter(self, node: ast.expr) -> ast.expr:
```

Comment:
This function does not use `self` and only converts the AST, so can be defined as an external utility (maybe `analyzers`?).